egs/chime4/s5_1ch/local/chime4_train_lms.sh

```diff
@@ -0,0 +1,163 @@
+else
+  echo "Training a $order-gram language model"
+  ngram-count -text $dir/train.gz -order $order \
+              -vocab $dir/vocab_5k.txt -unk -map-unk "<UNK>" \
+              -gt2min 1 -gt3min 1 -gt4min 2 -gt5min 2 \
+              -interpolate -kndiscount \
+              -lm $dir/lm_${lm_suffix}.arpa.gz
+fi
+echo "Checking validation perplexity of $order-gram language model"
+ngram -order $order -ppl $dir/valid.gz -lm $dir/lm_${lm_suffix}.arpa.gz
+# e.g. 5-gram perplexity:
+# file data/local/local_lm/valid.txt: 3280 sentences, 54239 words, 3 OOVs
+# 0 zeroprobs, logprob= -96775.5 ppl= 48.1486 ppl1= 60.8611
+
+# Next, create the corresponding FST and lang_test_* directory.
+echo "Preparing language models for test"
+tmpdir=data/local/lm_tmp
+lexicon=data/local/lang_tmp/lexiconp.txt
+mkdir -p $tmpdir
+
+test=data/lang_test_${lm_suffix}
+mkdir -p $test
+for f in phones.txt words.txt phones.txt L.fst L_disambig.fst \
+   phones; do
+  cp -r data/lang/$f $test
+done
+gunzip -c $dir/lm_${lm_suffix}.arpa.gz | \
+ utils/find_arpa_oovs.pl $test/words.txt  > $tmpdir/oovs_${lm_suffix}.txt
+
+# grep -v '<s> <s>' because the LM seems to have some strange and useless
+# stuff in it with multiple <s>'s in the history.  Encountered some other similar
+# things in a LM from Geoff.  Removing all "illegal" combinations of <s> and </s>,
+# which are supposed to occur only at being/end of utt.  These can cause
+# determinization failures of CLG [ends up being epsilon cycles].
+gunzip -c $dir/lm_${lm_suffix}.arpa.gz | \
+  grep -v '<s> <s>' | \
+  grep -v '</s> <s>' | \
+  grep -v '</s> </s>' | \
+  arpa2fst - | fstprint | \
+  utils/remove_oovs.pl $tmpdir/oovs_${lm_suffix}.txt | \
+  utils/eps2disambig.pl | utils/s2eps.pl | fstcompile --isymbols=$test/words.txt \
+    --osymbols=$test/words.txt  --keep_isymbols=false --keep_osymbols=false | \
+   fstrmepsilon | fstarcsort --sort_type=ilabel > $test/G.fst
+fstisstochastic $test/G.fst
+# The output is like:
+# 9.14233e-05 -0.259833
+# we do expect the first of these 2 numbers to be close to zero (the second is
+# nonzero because the backoff weights make the states sum to >1).
+# Because of the <s> fiasco for these particular LMs, the first number is not
+# as close to zero as it could be.
+
+# Everything below is only for diagnostic.
+# Checking that G has no cycles with empty words on them (e.g. <s>, </s>);
+# this might cause determinization failure of CLG.
+# #0 is treated as an empty word.
+mkdir -p $tmpdir/g
+awk '{if(NF==1){ printf("0 0 %s %s\n", $1,$1); }} END{print "0 0 #0 #0"; print "0";}' \
+    < "$lexicon"  >$tmpdir/g/select_empty.fst.txt
+fstcompile --isymbols=$test/words.txt --osymbols=$test/words.txt $tmpdir/g/select_empty.fst.txt | \
+ fstarcsort --sort_type=olabel | fstcompose - $test/G.fst > $tmpdir/g/empty_words.fst
+fstinfo $tmpdir/g/empty_words.fst | grep cyclic | grep -w 'y' &&
+  echo "Language model has cycles with empty words" && exit 1
+rm -r $tmpdir/g
+
```

Comment:
sorry guys, I didn't notice who was doing this PR, I assumed it was Shinji.

On Wed, Apr 27, 2016 at 3:35 PM, Shinji Watanabe <notifications@github.com>
wrote:

> In egs/chime4/s5_1ch/local/chime4_train_lms.sh
> <https://github.com/kaldi-asr/kaldi/pull/730#discussion_r61321918>:
>
> > +# Because of the <s> fiasco for these particular LMs, the first number is not
> > +# as close to zero as it could be.
> > +
> > +# Everything below is only for diagnostic.
> > +# Checking that G has no cycles with empty words on them (e.g. <s>, </s>);
> > +# this might cause determinization failure of CLG.
> > +# #0 is treated as an empty word.
> > +mkdir -p $tmpdir/g
> > +awk '{if(NF==1){ printf("0 0 %s %s\n", $1,$1); }} END{print "0 0 #0 #0"; print "0";}' \
> > +    < "$lexicon"  >$tmpdir/g/select_empty.fst.txt
> > +fstcompile --isymbols=$test/words.txt --osymbols=$test/words.txt $tmpdir/g/select_empty.fst.txt | \
> > + fstarcsort --sort_type=olabel | fstcompose - $test/G.fst > $tmpdir/g/empty_words.fst
> > +fstinfo $tmpdir/g/empty_words.fst | grep cyclic | grep -w 'y' &&
> > +  echo "Language model has cycles with empty words" && exit 1
> > +rm -r $tmpdir/g
> > +
>
> I just discussed it with @takaaki-hori <https://github.com/takaaki-hori>
> , He could handle it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/730/files/1bb204ec1d3cadd1d7760d9c6bb84d2cb5dd3357#r61321918>
>